modules/validation/helpers.go

```diff
@@ -91,3 +91,15 @@ func IsValidExternalTrackerURLFormat(uri string) bool {
 
 	return true
 }
+
+var (
+	validUsernamePattern   = regexp.MustCompile(`^[\da-zA-Z][-.\w]*$`)
+	invalidUsernamePattern = regexp.MustCompile(`[-._]{2,}|[-._]$`)
```

Comment:
`the__name` or `the_name_` could be fine?

Reply:
And IMO `--` could also be fine. At least, it's used by IDNA encoding.

If a user want to create a repo for a IDNA domain, it will be like `xn--abcd.com`

Reply:
I'm not sure, the only thing I can say is that GitHub doesn't support them and it works fine.

Actually, I don't really care what GitHub does, what I really want is a clear rule for username.

The previous rule was very clear: "alphanumeric, dash ('-'), underscore ('_') and dot ('.')", but it wasn't ideal. I am trying to fix it with "alphanumeric, dash ('-'), underscore ('_') and dot ('.'), start and end with alphanumeric, no consecutive non-alphanumerics". I think it’s clear enough.

However, if there are some exceptions, like "but consecutive dash or underscore is ok, and end with underscore is ok", it may not be too difficult to write such logic, but it's hard to describe the rule to users and explain why. Why "a--b" is ok and "a..b" isn't?

Reply:
I worry about that some users are using `__` and `--` for some special purpose (see the example above, like the IDNA encoding, etc, what if a user created a repo with IDNA encoding to store their website for a Unicode domain?), the strict  rule will affect them. Actually I also use `--` and `__` for some names in my code.


About the `..`, it's different, there is no well-known usage for such string, and it affects the cross-repo comparing.

Reply:
You may have misunderstood, I don't touch the rule of repo name, only username.

And I think your concern is right, but I think users can work around the problems.

Like:

- `https://github.com/dotnet` (for .Net)
- `https://github.com/cplusplus` (for C++)

Reply:
Hmm, right. Maybe for username it's fine.